### PR TITLE
DDP native mixed precision

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #92882

Implements native mixed precision support for DDP in a similar fashion to how it is enabled for FSDP. The implementation works as follows:

1. In DDP init, we save `_mp_param` and `_fp_param` variables to manage mixed precision parameter usage. In particular, _mp_param will represent the parameter in the reduced precision, while _fp_param will represent the param in regular precision. During forward/backward, we swap back and forth as needed. 
2. The root module gets a root pre-forward hook that kicks off copies to the reduced precision for all submodules. An event is recorded for each submodule to allow for waiting, as we run these asynchronously. 
3. Each module gets a pre-forward hook that waits on its corresponding event. note that modules might be reused during training, in this case the wait is only done for the first module. After this wait, the module's parameters are in reduced precision. 
4. In the pre-forward hook, we register a backward hook on the lower precision parameters in order to run reduced precision allreduce + parameter upcast. We can't rely on the Reducer's constructor setting up these hooks because the gradient is accumulated on the low precision param, so we need to register them ourselves.
5. In the backward pass, when the hook runs, we first run allreduce + divide in the reduced precision. Next, we upcast parameters and gradients back to fp32 asynchronously. We also queue a callback at the end of backward to wait on these upcasts so that the upcast is complete before optim.step() runs.
6. Parameters that don't require grad are also cast since they may be used in computation, they are upcast back in the final autograd callback.
7. DDP Ignored parameters are not touched.

Follow-ups:

1. Unify comm hooks and make it work with apply optimizer in backward
2. implement keep_low_precision_grads, 
3. allow BN, LN, or custom units to run in reduced precision, 
4. support for cast_forward_inputs
5. Unify certain APIs / helpers with FSDP where possible, such as for _cast_forward_inputs
6. Integrate this with replicate() API.
7. The order in which we kick off copies and wait for them is set by the iteration order of module.modules(), but this might not be how the modules are used in the actual training. In the worst case, the last module in module.modules() could be used first which would result in waiting for all copies unnecessarily. For static graphs, we should record the module execution order and copy / wait in this order. 
8. Entirely unused modules probably don't need to be cast.

Differential Revision: [D42515803](https://our.internmc.facebook.com/intern/diff/D42515803/)

### CODE REVIEW

File: torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        # Parameters that don't require grad still needed to be casted since
+        # they may participate in computation. However, they would not be recast
+        # by hook above as they don't have a grad hook installed, so cast them
+        # back here.
+        for n, p in ddp_weakref().module.named_parameters():
+            if hasattr(p, '_ddp_mp_hook_state'):
+                p._ddp_mp_hook_state[1].remove()
+                delattr(p, '_ddp_mp_hook_state')
+            if not p.requires_grad and not hasattr(p, '_ddp_ignored'):
+                p.data = p._fp_param
```
Comment (zhaojuanmao) [inline torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py:70]: the upcast has already been done in above codes?

**Review by zhaojuanmao — COMMENTED** [review]
thanks for putting them together!! left some comments, also would you please use ".data" as less as possible?

File: torch/nn/parallel/distributed.py [hunk 2]
```diff
@@ -44,6 +53,105 @@
...
+    param_dtype: Optional[torch.dtype] = None
+    reduce_dtype: Optional[torch.dtype] = None
+    buffer_dtype: Optional[torch.dtype] = None
+    # TODO (rohan-varma): keep_low_precision_grads: bool = False
+    cast_root_forward_inputs: bool = True
+    # TODO (rohan-varma): APIs to allow users to run batchnorm and layernorm
+    # in full precision. For DDP, this can be implemented by not performing the
+    # parameter cast for BN and LN units.
+    keep_batchnorm_fp32: bool = True
+    keep_layernorm_fp32: bool = True
```
[the remark this answers is not in the document]
Comment (zhaojuanmao) [reply torch/nn/parallel/distributed.py:102]: also did not see any code to handle keep_batchnorm_fp32/keep_layernorm_fp32 right now

File: torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    ):
+        # Cast bucket tensor to reduce_dtype
+        bucket.set_buffer(bucket.buffer().to(ddp_weakref().mixed_precision.reduce_dtype))
+    fut = reducer._run_allreduce_hook(bucket)
+    ret_fut = torch.futures.Future()
+    stream = hook_state.upcast_stream
+    with torch.cuda.stream(stream):
+        fut.wait()
+        bucket.buffer().div_(process_group.size())
+        ret_fut.set_result(bucket.buffer())
```
Comment (zhaojuanmao) [inline torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py:43]: the allreduced bucket will be used to update grads in "finalize_bucket_dense()" later on, so the below upcast codes seem not working? the upcast codes possible should be done after "finalize_bucket_dense()"?

File: torch/nn/parallel/distributed.py [hunk 4]
```diff
@@ -752,6 +902,79 @@ def _setup_in_backward_optimizers(self):
...
+                continue
+            # We need to register autograd hook here instead of DDP's ctor
+            # since we're working with the low precision param. Register them
+            # via obtaining the gradient accumulator.
+            tmp = p.expand_as(p)
+            grad_acc = tmp.grad_fn.next_functions[0][0]
+
+            hook = grad_acc.register_hook(
+                functools.partial(self._fire_reducer_autograd_hook, p._idx)
+            )
```
[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/nn/parallel/distributed.py:975]: They are not deregistered. If we'd like, we could have an API to de-register them if mixed precision is enabled.

File: torch/nn/parallel/distributed.py [hunk 5]
```diff
@@ -44,6 +53,105 @@
...
+        happens in ``param_dtype`` if specified or the original parameter dtype
+        otherwise. For example, ``MixedPrecision(param_dtype=torch.float16)``
+        would result in communication ocurring in fp16.
+    """
+
+    param_dtype: Optional[torch.dtype] = None
+    reduce_dtype: Optional[torch.dtype] = None
+    buffer_dtype: Optional[torch.dtype] = None
+    # TODO (rohan-varma): keep_low_precision_grads: bool = False
+    cast_root_forward_inputs: bool = True
```
[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/nn/parallel/distributed.py:97]: that's true, unless the first unit or something is hardcoded to run in fp32. but we can add this on a use case basis.

File: torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+@dataclass
+class _AllreduceUpcastHookState:
+    """
+    State to manage DDP mixed precision in backward / gradient communication.
+    This contains a weakref to the DDP module for access to reducer and process
+    group, and a stream to run parameter and gradient upcasts.
+    """
+    ddp_weakref: Any
+    upcast_stream: torch.cuda.Stream
+    upcast_params_and_grads: bool
```
[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py:17]: I can take this out for now, but the reason for this flag is that eventually we'll unify this state with the optimizer overlap comm hook, in which case the unified code would not always upcast params and grads

File: torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py (hunk 1 above)
Comment (rohan-varma) [reply torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py:70]: This is for params that don't require grad. They don't get the above code run since their autograd hooks do not fire, but we still need to upcast them as they were downcast before the forward (we need to do that since they may participate in computation)

File: torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py (hunk 3 above)
Comment (rohan-varma) [reply torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py:43]: hmm this is a good point, let me look into it

Comment (rohan-varma) [reply torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py:43]: The upcast code actually runs before `finalize_bucket_dense`: 

1. This code is kicked off in the comm hook callback, and run on a separate stream. This stream is waited on by the below queue_callback to the autograd engine. 
2. C++ finalize_backward also happens as an autograd engine final callback. However it runs _after_ the python callback, since we insert the python callback first, which is why the upcast code here works as expected. 
3. It is not great to rely on autograd engine callback running order. We can look into consolidating and inserting only a single callback always for DDP.

File: torch/nn/parallel/distributed.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/nn/parallel/distributed.py:102]: Will add this as follow up work.

File: torch/nn/parallel/distributed.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/nn/parallel/distributed.py:97]: Removing this for now.

File: torch/nn/parallel/distributed.py [hunk 7]
```diff
@@ -1109,9 +1339,30 @@ def _run_ddp_forward(self, *inputs, **kwargs):
...
             with self._inside_ddp_forward():
-                return module_to_run(*inputs[0], **kwargs[0])  # type: ignore[index]
+                return module_to_run(*args, **kwargs)  # type: ignore[index]
         else:
+            # Cast inputs to reduced precision if needed.
+            if (
+                self.mixed_precision is not None
+                and self.mixed_precision.cast_root_forward_inputs
+            ):
+                inputs, kwargs = _cast_forward_inputs(
```
[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/nn/parallel/distributed.py:1361]: We could investigate this but that would mean that in `device_ids` path (device_id is usually provided if GPU training), we would cast before the CPU -> GPU copy. here we're sticking with FSDP pattern of casting after copy to GPU, if we want to merge this, we can evaluate the perf impact as follow up work?

File: torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    with torch.cuda.stream(stream):
+        fut.wait()
+        bucket.buffer().div_(process_group.size())
+        ret_fut.set_result(bucket.buffer())
+
+        if hook_state.upcast_params_and_grads:
+            # Upcast parameters and gradients so optimizer step can run in fp32.
+            params, grads = bucket.parameters(), bucket.gradients()
+            for p, g in zip(params, grads):
+                p.data = p._fp_param
```
[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py:49]: Yes, great cal out!

File: torch/nn/parallel/distributed.py [hunk 9]
```diff
@@ -44,6 +53,105 @@
...
+    """
+    Creates and frees storage for the mixed precision parameters.
+    """
+    for param in root_module.parameters():
+        # Do not setup mixed precision for DDP ignored parameters.
+        if hasattr(param, '_ddp_ignored') and param._ddp_ignored:
+            continue
+
+        if not hasattr(param, '_mp_param'):
+            param._mp_param = torch.zeros_like(
```
Comment (awgu) [inline torch/nn/parallel/distributed.py:126]: nit: I wonder if `_mp_param` is a good name. Maybe it should be `_lp_param` for "low precision parameter". Mixed precision describes the overall methodology of using both low and full precision.

File: torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+        ret_fut.set_result(bucket.buffer())
+
+        # Upcast parameters and gradients so optimizer step can run in fp32.
+        params, grads = bucket.parameters(), bucket.gradients()
+        for p, g in zip(params, grads):
+            p.data = p._fp_param
+            # free storage for mp param as it will be allocated again in next
+            # forward pass.
+            _free_storage(p._mp_param)
+            p.grad.data = p.grad.to(p.data.dtype)
```
Comment (zhaojuanmao) [inline torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py:53]: nit: p.grad = p.grad.to(p.dtype), avoid using .data

File: torch/nn/parallel/distributed.py [hunk 11]
```diff
@@ -43,6 +49,97 @@
...
 logger = logging.getLogger(__name__)
 
+@dataclass
+class _MixedPrecision:
+    """
+    This configures DDP-native mixed precision training.
+
+    Attributes:
+        param_dtype (torch.dtype): This specifies the dtype for model
+            parameters, inputs (when ``cast_forward_inputs`` is set to
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:59]: nit: remove 'cast_forward_inputs' related comments as it is not used anymore

File: torch/nn/parallel/distributed.py [hunk 12]
```diff
@@ -43,6 +49,97 @@
...
+
+def _cast_buffers(mixed_precision_config, root_module):
+    """
+    Casts buffers to the given ``buffer_dtype``.
+    """
+    for buf in root_module.buffers():
+        if hasattr(buf, '_ddp_ignored') and buf._ddp_ignored:
+            continue
+
+        buf.data = buf.to(dtype=mixed_precision_config.buffer_dtype)
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:102]: nit: could we do 'buf = buf.to()'?

File: torch/nn/parallel/distributed.py [hunk 13]
```diff
@@ -43,6 +49,97 @@
...
+    Creates and frees storage for the mixed precision parameters.
+    """
+    for param in root_module.parameters():
+        # Do not setup mixed precision for DDP ignored parameters.
+        if hasattr(param, '_ddp_ignored') and param._ddp_ignored:
+            continue
+
+        if not hasattr(param, '_mp_param'):
+            param._mp_param = torch.zeros_like(
+                param.data,
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:115]: nit: remove .data here

File: torch/nn/parallel/distributed.py [hunk 14]
```diff
@@ -43,6 +49,97 @@
...
+                device=param.device,
+                dtype=mixed_precision_config.param_dtype,
+                requires_grad=param.requires_grad,
+            )
+            _free_storage(param._mp_param)
+            # _fp_param will point to the full precision param so it can be switched
+            # back to at the end of forward / backward.
+            param._fp_param = param.data
+
+def _cast_forward_inputs(
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:125]: nit: move to util helper

File: torch/nn/parallel/distributed.py [hunk 15]
```diff
@@ -752,6 +902,79 @@ def _setup_in_backward_optimizers(self):
...
+        with torch.cuda.stream(self._mp_stream):
+            for submodule in self.module.modules():
+                for param in submodule.parameters(recurse=False):
+                    # Do not cast DDP ignored parameters.
+                    if hasattr(param, '_ddp_ignored') and param._ddp_ignored:
+                        continue
+                    _alloc_storage(param._mp_param, param.data.size())
+                    # copy() implicitly casts to low precision
+                    with torch.no_grad():
+                        param._mp_param.copy_(param.data)
```
[the remark this answers is not in the document]
Comment (zhaojuanmao) [reply torch/nn/parallel/distributed.py:918]: nit: remove .data

File: torch/nn/parallel/distributed.py [hunk 16]
```diff
@@ -745,6 +883,90 @@ def _setup_in_backward_optimizers(self):
...
+                        # accumulation case, accumulated grads can be in fp32
+                        # which can cause errors when running DDP backwards due
+                        # to mismatched incoming and accumulated gradient types.
+                        # So we manually cast the accumulated grad down for now,
+                        # in the future we may shift to FSDP style gradient
+                        # accumulation management where the accumulated gradient
+                        # is saved and .grad field is set to None, bypassing
+                        # this issue.
+                        if param.grad is not None:
+                            param.grad.data = param.grad.to(self.mixed_precision.param_dtype)
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:929]: nit: remove .data?

**Review by zhaojuanmao — COMMENTED** [review]
overall looks good, since the code path is gated, it should be safe to land.  Left some comments to remove .data usage, can be done in follow up PR. 

Another important thing, I just realized that buckets and grads are mostly always detached because of the down cast and up cast in every iteration's forward, we should make grads point to buckets at the end of forward to avoid coping btw buckets and grads, otherwise it may have performance regression for some workloads. This can be added in the follow up PR though.

**Review by zhaojuanmao — APPROVED** [review]

File: torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py (hunk 10 above)
Comment (rohan-varma) [reply torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py:53]: discussed offline to leave as is, otherwise when set_to_none=False, the gradients could potentially become unlinked to buckets here

File: torch/nn/parallel/distributed.py (hunk 12 above)
Comment (rohan-varma) [reply torch/nn/parallel/distributed.py:102]: this doesn't work since it will just point buf to the fp16 tensor, but the buffer .buffers() won't be updated. FSDP does the same thing.

If we'd like to avoid .data usage here, we could de-register the float32 buffer, and register a fp16 buffer.

File: torch/nn/parallel/distributed.py (hunk 16 above)
Comment (rohan-varma) [reply torch/nn/parallel/distributed.py:931]: similar to above, this would de-link the pointers for grad_as_bucket_view (when set_grad_none=False), so decided to keep as is for now.
